### PR TITLE
[FLINK-12458][network] Introduce PartitionConnectionException for unreachable producer

### PR DESCRIPTION
**Pull request by zhijiangW**
## What is the purpose of the change

* If the consumer can not establish a connection to remote task executor while requesting remote subpartition, which might indicate the remote task executor is not reachable. We could wrap this connection exception into new proposed `PartitionConnectionException` which also extends `PartitionException`, then the job master would decide whether to restart the upstream region to re-producer partition data.*

## Brief change log

  - *Introduce new `PartitionConnectionException` extends `PartitionException.*
  - *Wrap any exceptions into `PartitionConnectionException` while creating partition request client during `RemoteInputChannel#requestSubpartition`*

## Verifying this change

  - *Added new tests in `RemoteInputChannelTest` for verifying the `PartitionConnectionException` while requesting sub partition.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zhijiangW** [comment]
@flinkbot attention @tillrohrmann

**Comment by tillrohrmann** [comment]
The PR does not compile @zhijiangW.

**Comment by zhijiangW** [comment]
Thanks for kindly reminding, it is because of unused import checkstyle. I already fixed the issue.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java [hunk 1]
```diff
@@ -159,11 +159,15 @@ void assignExclusiveSegments(List<MemorySegment> segments) {
...
 	@Override
-	public void requestSubpartition(int subpartitionIndex) throws IOException, InterruptedException {
+	public void requestSubpartition(int subpartitionIndex) throws IOException {
 		if (partitionRequestClient == null) {
 			// Create a client and request the partition
-			partitionRequestClient = connectionManager
-				.createPartitionRequestClient(connectionId);
+			try {
+				partitionRequestClient = connectionManager.createPartitionRequestClient(connectionId);
+			} catch (Throwable t) {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java]: I think we should only catch `IOException` here. If a different exception is thrown, this indicates a serious Flink problem and we should best fail the `TaskExecutor` process.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java [hunk 2]
```diff
@@ -1047,6 +1047,22 @@ public void testPartitionNotFoundExceptionWhileRetriggeringRequest() throws Exce
...
+	 * Tests that any exceptions thrown by {@link ConnectionManager#createPartitionRequestClient(ConnectionID)}
+	 * would be wrapped into {@link PartitionConnectionException} during
+	 * {@link RemoteInputChannel#requestSubpartition(int)}.
+	 */
+	@Test
+	public void testPartitionConnectionExceptionWhileRequestingPartition() throws Exception {
+		final RemoteInputChannel inputChannel = InputChannelTestUtils.createRemoteInputChannel(
+			createSingleInputGate(1), 0, new TestingExceptionConnectionManager());
+		try {
+			inputChannel.requestSubpartition(0);
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java]: fail is missing

Comment (zhijiangW) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java]: fail is not missing on my side?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java (hunk 1 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java]: Only the `InterruptedException` might be thrown during `future.waitForChannel()`. If we only catch the `IOException`, then for other cases we could keep the previous behavior to throw the raw exception directly?

**Comment by zhijiangW** [comment]
@tillrohrmann , thanks for the review!
If I understood correctly, it does not need to rebase master on my side?

**Comment by tillrohrmann** [comment]
Yes, I pushed the changes to your branch. No need for any actions from your side.

**Pull request closed without merging**